### PR TITLE
Add support for saved cards

### PR DESCRIPTION
**Pull request by lchrusciel**
API: 
image
image
UI: 
image

image

### CODE REVIEW

File: config/services/event_listener.php [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+use CommerceWeavers\SyliusTpayPlugin\EventListener\AddCreditCardToAccountMenuEventListener;
+use Sylius\Bundle\ShopBundle\Menu\AccountMenuBuilder;
+
+return static function(ContainerConfigurator $container): void {
+    $services = $container->services();
+
+    $services->set(AddCreditCardToAccountMenuEventListener::class)
+        ->tag('kernel.event_listener', ['event' => AccountMenuBuilder::EVENT_NAME, 'method' => '__invoke'])
+
+    ;
```
Comment (arti0090) [inline config/services/event_listener.php:18]:
```suggestion
    $services->set(AddCreditCardToAccountMenuEventListener::class)
        ->tag('kernel.event_listener', ['event' => AccountMenuBuilder::EVENT_NAME, 'method' => '__invoke'])
    ;
```

File: src/Entity/CreditCard.php [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    private ?string $token = null;
+    private ?string $brand = null;
+    private ?string $tail = null;
+
+    private ?\DateTimeInterface $expirationDate = null;
+
+    private ?CustomerInterface $customer = null;
+
+    private ?ChannelInterface $channel = null;
```
Comment (arti0090) [inline src/Entity/CreditCard.php:22]: Why some of the fields separated by blank line and some not?

File: src/Entity/CreditCard.php [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+    public function getTail(): ?string
+    {
+        return $this->tail;
+    }
+
+    public function setTail(?string $tail): void
+    {
+        $this->tail = $tail;
+    }
```
Comment (arti0090) [inline src/Entity/CreditCard.php:57]: missing blank line

File: src/Payum/Request/Api/SaveCreditCard.php [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+class SaveCreditCard extends Generic
+{
+    public function __construct(
+        mixed $model,
+        public readonly string $cardToken,
+        public readonly string $cardBrand,
+        public readonly string $cardTail,
+        public readonly string $tokenExpiryDate,
+
```
Comment (arti0090) [inline src/Payum/Request/Api/SaveCreditCard.php:17]: blank space not needed :)

File: src/Payum/Request/Api/SaveCreditCard.php [hunk 5]
```diff
@@ -0,0 +1,21 @@
+<?php
+
+declare(strict_types=1);
+
+namespace CommerceWeavers\SyliusTpayPlugin\Payum\Request\Api;
+
+use Payum\Core\Request\Generic;
+
+class SaveCreditCard extends Generic
```
Comment (arti0090) [inline src/Payum/Request/Api/SaveCreditCard.php:9]: I see in others Generics we are using getters to retrieve values, may I ask why you didn't?

File: translations/messages.en.yaml [hunk 6]
```diff
@@ -19,6 +19,11 @@ commerce_weavers_sylius_tpay:
                 redirect: 'Redirect to Tpay'
                 pay_by_link: 'Redirect to bank'
     shop:
+        credit_cards: 'Credit cards'
+        credit_card:
+            brand: 'Brand'
+            expiration_date: 'Expiration date'
+            tail: 'Tail'
```
Comment (arti0090) [inline translations/messages.en.yaml:26]: These translation keys are missing for PL

File: config/services/payum/action.php [hunk 7]
```diff
@@ -93,7 +94,18 @@
         ->tag('payum.action', ['factory' => TpayGatewayFactory::NAME, 'alias' => 'cw.tpay.notify'])
     ;
 
+    $services->set(SaveCreditCardAction::class)
+        ->args([
+            service('commerce_weavers_sylius_tpay.factory.credit_card'),
+            service('commerce_weavers_sylius_tpay.repository.credit_card'),
+        ])
+        ->tag('payum.action', ['factory' => TpayGatewayFactory::NAME, 'alias' => 'cw.tpay.credit_card'])
```
Comment (coldic3) [inline config/services/payum/action.php:102]:
```suggestion
        ->tag('payum.action', ['factory' => TpayGatewayFactory::NAME, 'alias' => 'cw.tpay.save_credit_card'])
```

File: config/serialization/Pay.xml [hunk 8]
```diff
@@ -29,6 +29,12 @@
         <attribute name="encodedCardData">
             <group>commerce_weavers_sylius_tpay:shop:order:pay</group>
         </attribute>
+        <attribute name="saveCard">
+            <group>commerce_weavers_sylius_tpay:shop:order:pay</group>
+        </attribute>
+        <attribute name="savedCardId">
+            <group>commerce_weavers_sylius_tpay:shop:order:pay</group>
+        </attribute>
         <attribute name="tpayChannelId">
```
Comment (coldic3) [inline config/serialization/Pay.xml:38]: I wonder, do we need `saveCard` and `saveCardId`. If `saveCardId` is null then we know `saveCard` is false. Or I am missing something?

File: migrations/Version20241030214842.php [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace CommerceWeaversSyliusTpayMigrations;
+
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
+final class Version20241030214842 extends AbstractMigration
```
Comment (coldic3) [inline migrations/Version20241030214842.php:13]: Let's merge it into the migration above.

File: phpunit.xml.dist [hunk 10]
```diff
@@ -37,6 +37,9 @@
         <env name="APP_ENV" value="test"/>
         <env name="SHELL_VERBOSITY" value="-1" />
         <env name="TPAY_API_URL" value="http://localhost:4000" />
+        <env name="JWT_SECRET_KEY" value="%kernel.project_dir%/config/jwt/private-test.pem" />
```
Comment (coldic3) [inline phpunit.xml.dist:40]: Since this is a test application, I believe we can use the same JWT keys across all environments.

Comment (lchrusciel) [reply phpunit.xml.dist:40]: I had some issues to run it locally, if the build will be green I would prefer to leave it for now

File: config/serialization/Pay.xml (hunk 8 above)
Comment (lchrusciel) [reply config/serialization/Pay.xml:38]: it is `saveCard` and `save**D**CardId`. One is a boolean flag to save encodedData other contains ID of credit card from customer. Boolean flag have sense only when send together with `encodedData`

File: config/services/form.php [hunk 11]
```diff
@@ -70,4 +76,13 @@
     ;
 
     $services->set('commerce_weavers_sylius_tpay.form.event_listener.remove_unnecessary_payment_details_fields', RemoveUnnecessaryPaymentDetailsFieldsListener::class);
+
+    $services
+        ->set('commerce_weavers_sylius_tpay.form.event_listener.add_saved_credit_cards', AddSavedCreditCardsListener::class)
```
Comment (lchrusciel) [inline config/services/form.php:81]: Dead code, should be removed

File: config/services/payum/mapper.php [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+use CommerceWeavers\SyliusTpayPlugin\Payum\Action\Api\NotifyAction;
+use CommerceWeavers\SyliusTpayPlugin\Payum\Action\Api\PayWithCardAction;
+use CommerceWeavers\SyliusTpayPlugin\Payum\Action\Api\SaveCreditCardAction;
+use CommerceWeavers\SyliusTpayPlugin\Payum\Action\CaptureAction;
+use CommerceWeavers\SyliusTpayPlugin\Payum\Action\GetStatusAction;
+use CommerceWeavers\SyliusTpayPlugin\Payum\Action\RefundAction;
+use CommerceWeavers\SyliusTpayPlugin\Payum\Action\ResolveNextRouteAction;
+use CommerceWeavers\SyliusTpayPlugin\Payum\Factory\TpayGatewayFactory;
+use CommerceWeavers\SyliusTpayPlugin\Payum\Mapper\PayWithCardActionPayloadMapper;
+use CommerceWeavers\SyliusTpayPlugin\Payum\Mapper\PayWithCardActionPayloadMapperInterface;
```
Comment (lchrusciel) [inline config/services/payum/mapper.php:25]: should be reduced

File: src/Api/Command/PayByCardHandler.php [hunk 13]
```diff
@@ -15,16 +15,17 @@ public function __invoke(PayByCard $command): PayResult
...
 
-        $this->setTransactionData($payment, $command->encodedCardData);
+        $this->setTransactionData($payment, $command->encodedCardData, $command->saveCard);
         $this->createTransactionProcessor->process($payment);
 
         return $this->createResultFrom($payment);
     }
 
-    private function setTransactionData(PaymentInterface $payment, string $encodedCardData): void
+    private function setTransactionData(PaymentInterface $payment, string $encodedCardData, bool $saveCard = false): void
```
Comment (lchrusciel) [inline src/Api/Command/PayByCardHandler.php:24]: does not need default value

File: src/Form/EventListener/AddSavedCreditCardsListener.php [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+use Sylius\Component\Core\Model\PaymentInterface;
+use Sylius\Component\Core\Model\ShopUserInterface;
+use Symfony\Component\Form\Extension\Core\Type\ChoiceType;
+use Symfony\Component\Form\FormEvent;
+use Symfony\Component\Form\FormInterface;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+use Symfony\Component\Translation\TranslatableMessage;
+use Symfony\Contracts\Translation\TranslatorInterface;
+
+final class AddSavedCreditCardsListener
```
Comment (lchrusciel) [inline src/Form/EventListener/AddSavedCreditCardsListener.php:19]: dead code, can be removed

File: src/Form/Type/TpayPaymentDetailsType.php [hunk 15]
```diff
@@ -101,5 +115,61 @@ public function buildForm(FormBuilderInterface $builder, array $options): void
...
+                ->add(
+                    'use_saved_credit_card',
+                    ChoiceType::class,
+                    [
+                        'label' => 'commerce_weavers_sylius_tpay.shop.order_summary.card.use_saved_credit_card.label',
+                        'placeholder' => new TranslatableMessage('commerce_weavers_sylius_tpay.shop.credit_card.use_new_card'),
+                        'required' => false,
+                        'choices' => $choices,
+                    ],
+                )
```
Comment (lchrusciel) [inline src/Form/Type/TpayPaymentDetailsType.php:171]: could be extracted to separate form type

File: src/Payum/Mapper/PayWithCardActionPayloadMapper.php [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+     */
+    public function getPayload(PaymentDetails $paymentDetails): array
+    {
+        $payload = [
+            'groupId' => PayGroup::CARD,
+        ];
+
+        if ($paymentDetails->getUseSavedCreditCard() !== null) {
+            /** @var CreditCardInterface|null $creditCard */
+            $creditCard = $this->creditCardRepository->find($paymentDetails->getUseSavedCreditCard());
```
Comment (lchrusciel) [inline src/Payum/Mapper/PayWithCardActionPayloadMapper.php:30]: we have to take into account channel and customer with this query

File: templates/shop/payment/_card.html.twig [hunk 17]
```diff
@@ -18,7 +31,7 @@
                 </div>
             </div>
             <div class="fields">
-                <div class="six wide field" data-tpay-field>
+                <div class="three wide field" data-tpay-field>
```
Comment (lchrusciel) [inline templates/shop/payment/_card.html.twig:34]: this value should be conditional based on the latter if statement

File: tests/Api/DataFixtures/shop/paying_for_orders_by_card/cart.yml [hunk 18]
```diff
@@ -5,6 +5,16 @@ Sylius\Component\Core\Model\Customer:
...
 
+Sylius\Component\Core\Model\ShopUser:
+    user_john_doe:
+        plainPassword: '123pa\\$\\$word'
+        roles: [ROLE_USER]
+        enabled: true
+        customer: '@customer_john_doe'
+        username: 'sylius@example.com'
+        usernameCanonical: 'sylius@example.com'
+
```
Comment (lchrusciel) [inline tests/Api/DataFixtures/shop/paying_for_orders_by_card/cart.yml:16]:
```suggestion
```

File: src/Payum/Mapper/PayWithCardActionPayloadMapper.php (hunk 16 above)
Comment (lchrusciel) [reply src/Payum/Mapper/PayWithCardActionPayloadMapper.php:30]: https://github.com/CommerceWeavers/SyliusTpayPlugin/pull/152

**Pull request merged**